dask/array/core.py

```diff
@@ -3480,9 +3485,8 @@ def from_array(
         asarray = not hasattr(x, "__array_function__")
 
     previous_chunks = getattr(x, "chunks", None)
-
     chunks = normalize_chunks(
-        chunks, x.shape, dtype=x.dtype, previous_chunks=previous_chunks
+        chunks, x.shape, dtype=x.dtype, previous_chunks=previous_chunks  # type: ignore[union-attr]
```

Comment:
Hi @benrutter - I saw your comments on #11047, I'm hoping to get your opinion here.

I'm getting a lot (~60) of errors from mypy triggered by my type annotations. 

In this case,  mypy is warning that `x` may not have the `shape` or `dtype` attributes. This is true given the docstring-specified annotation of `ArrayLike`, and the code so far. (For example `x` could be a set.)

My question is - do you think it would be better to ship the type annotations with ignore statements for now? Or do we perhaps want to break this into a couple PRs, and fix type conflicts function by function?

Reply:
Hmm, that's a good question! I should say I've only made a few small contributions to dask so might not be the best person to steer on strategy for type hinting.

My instinct is that mypy is right and that the docstring specified "ArrayLike" is technically wrong. Like you said, passing in a set would be correct based on the types hinted but would give an error.

I wonder if it would be more accurate to specify the type as an NDArray from numpy.typing? That seems more accurate with what's actually being expected to get passed in to the function? What do you think?

Reply:
(I said more accurate, but I guess I also mwan "avoid throwing loads of mypy errors) 😅

Reply:
Gotcha!

I think `ArrayLike` is actually correct. For example you can see on (original) line 3476 that the function expects types that correspond to `ArrayLike`, including `np.ScalarType`. So annotating `x` as `ndarray` might be misleading.

As an aside, `ArrayLike` is more broad than I realized:

https://numpy.org/devdocs/reference/typing.html#numpy.typing.ArrayLike

Replacing the condition on line 3476 with something like `not isinstance(x, (ndarray, Array))` solves the mypy error. It's just that adding this kind of fix for every mypy error changes the scope and purpose of this PR.

I'm leaning toward just keeping the `ignore` statements. That way the issue-raiser gets the functionality they need and this PR doesn't get bogged down pulling at this thread.

Thanks for the feedback!

Reply:
That makes sense! Don't think I was understanding the flow of the variable x to be honest 😅

Adding type ignores sounds like it at least gives the user type hints even if they aren't being used internally.

I'm mainly just asking this for my own learning since you've soent a bunch of time working through the code, but us what's hapenening something like this:

- x starts as ArrayLike
- the below line narrows it to be an ndarray?
```python
if isinstance(x, (list, tuple, memoryview) + np.ScalarType):
  x = np.array(x)
```

(if that is the case, I guess mypy would want some new variable with a different more specific type? or a bunch of casts, but you're probably right that that could be a rabbit hole)

Reply:
Yeah it's a bit confusing.

*If* the caller passes an `ArrayLike`, `x` starts as `ArrayLike`.

The line you quote only checks for the specified types. Some types will pass through But `ArrayLike` encompasses many more types. For example, I believe `x` could be passed as a string. I think `np.array()` accepts strings, making them `ArrayLike`. 

So a string would pass a static type check, but not be caught by this `if` statement. `from_array` would end up throwing an error on original line 3485, when we try to access the string's `shape`. (This is what mypy was complaining about.)

It may be worth adding a line to enforce the `ArrayLike` type, and cast everything that passes the check to an `ndarray`. Though my sense is that correcting all such errors in the code is going to be pulling on a long thread. After all, nobody is complaining about their string causing `from_array` to error!

Unless folks object I'm leaning toward keeping the `ignore` statements so we can be sure to deliver the feature someone is actually asking for.

Reply:
For what it's worth I think that's a good call!